pkg/cloud/vsphere/actuators/cluster/actuator.go

```diff
@@ -115,7 +142,7 @@ func (a *Actuator) GetIP(cluster *clusterv1.Cluster, machine *clusterv1.Machine)
 
 // GetKubeConfig returns the contents of a Kubernetes configuration file that
 // may be used to access the cluster.
-func (a *Actuator) GetKubeConfig(cluster *clusterv1.Cluster, machine *clusterv1.Machine) (string, error) {
+func (a *Actuator) GetKubeConfig(cluster *clusterv1.Cluster, _ *clusterv1.Machine) (string, error) {
```

Comment:
Assumed it's safe to ignore the machine resource here since cluster context provides enough information to generate the kubeconfig.

Reply:
Hi @andrewsykim,

That's not accurate. Please see this squashed commit comment from bd46a0bf5fc6c2c88e5d8b1f6bf254bb3d4309ab:

> This patch allows a machine to return a control plane endpoint to satisfy the ProviderDeployer interface. Please see the Slack conversation https://kubernetes.slack.com/archives/C8TSNPY4T/p1560696064368600 for more information.

Reply:
The control plane endpoints are fetched in the cluster status though? I tested this change as well and it works as expected.

Reply:
The params that GetKubeConfig checks in the context only checks fields set by the initial cluster context

Reply:
So `GetKubeConfig` [invokes](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/f5893540c99c40f46f9520070d6964b463705ea1/pkg/cloud/vsphere/services/kubeclient/kubeclient.go#L49) `ctx.ControlPlaneEndpoint()`.

When `ctx KubeContext`, as passed into `GetKubeConfig`, is a `MachineContext`, [this](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/f5893540c99c40f46f9520070d6964b463705ea1/pkg/cloud/vsphere/context/machine_context.go#L163-L182) is what gets called:

```golang
// ControlPlaneEndpoint returns the control plane endpoint for the cluster.
// This function first attempts to retrieve the control plane endpoint with
// ClusterContext.ControlPlaneEndpoint.
// If no endpoint is returned then this machine's IP address is used as the
// control plane endpoint if the machine is a control plane node.
// Otherwise an error is returned.
func (c *MachineContext) ControlPlaneEndpoint() (string, error) {


	if controlPlaneEndpoint, _ := c.ClusterContext.ControlPlaneEndpoint(); controlPlaneEndpoint != "" {
		return controlPlaneEndpoint, nil
	}


	ipAddr := c.IPAddr()
	if ipAddr == "" || !c.HasControlPlaneRole() {
		return "", errors.New("unable to get control plane endpoint")
	}
	controlPlaneEndpoint := net.JoinHostPort(ipAddr, strconv.Itoa(int(c.BindPort())))
	c.Logger.V(2).Info("got control plane endpoint from machine", "control-plane-endpoint", controlPlaneEndpoint)
	return controlPlaneEndpoint, nil
}
```

So what actually happens is the Cluster's fields are checked first, and if nothing is returned, the IP from the machine is used to build the control plane endpoint.

Reply:
I submitted the previous comment too soon.

> So what actually happens is the Cluster's fields are checked first, and if nothing is returned, the IP from the machine is used to build the control plane endpoint.

Why is this important? Well, because the Cluster's fields aren't even updated until the cluster is patched. Calling the new `reconcileKubeConfig` function before the first time a cluster is patched won't actually work.

Here's what I recommend:

* Move the logic that assigns a cluster's APIEndpoints out of the `ClusterContext` patch call and into `reconcileControlPlaneEndpoint` in the cluster actuator. Put that above the `reconcileKubeConfig` call. If the `reconcileControlPlaneEndpoint` fails, then requeue the cluster. Once `reconcileControlPlaneEndpoint` works you can then call `reconcileKubeConfig`.

Also, please leave the `GetKubeConfig` function the way it was as it is consistent with the deployer, and the two context structs are designed to work with it.

Reply:
Ah I see.. and if we use cluster context then a random machine in the control plane will be used. Any ideas on how to pass a machine context without any references to machines from `Reconcile`? We can list machines and pick one but that wouldn't be much different from what cluster context does with it's `ControlPlaneEndpoint`

Reply:
> then a random machine in the control plane will be used

Not random -- the oldest control plane node is always used.

Reply:
I know we resolved this via Zoom, but I wanted to clarify the above in case someone else reads this :)